<strong>Week-3 Assignment</strong>

<strong>Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto</strong>

In [3]:
import pandas as pd

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [32]:
path='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [33]:
d=pd.read_html(path,flavor='bs4',skiprows=2,header=None)

In [34]:
d

[     M2A      Not assigned                                         Unnamed: 2
 0    M3A        North York                                          Parkwoods
 1    M4A        North York                                   Victoria Village
 2    M5A  Downtown Toronto                         Regent Park / Harbourfront
 3    M6A        North York                  Lawrence Manor / Lawrence Heights
 4    M7A  Downtown Toronto       Queen's Park / Ontario Provincial Government
 5    M8A      Not assigned                                                NaN
 6    M9A         Etobicoke                                   Islington Avenue
 7    M1B       Scarborough                                    Malvern / Rouge
 8    M2B      Not assigned                                                NaN
 9    M3B        North York                                          Don Mills
 10   M4B         East York                   Parkview Hill / Woodbine Gardens
 11   M5B  Downtown Toronto                         

In [35]:
type(d)

list

In [36]:
df=d[0]

df.head()

M2A      Not assigned                                    Unnamed: 2
0  M3A        North York                                     Parkwoods
1  M4A        North York                              Victoria Village
2  M5A  Downtown Toronto                    Regent Park / Harbourfront
3  M6A        North York             Lawrence Manor / Lawrence Heights
4  M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [37]:
type(df)

pandas.core.frame.DataFrame

In [38]:
df.columns

Index(['M2A', 'Not assigned', 'Unnamed: 2'], dtype='object')

In [39]:
cols=['Postal Code','Borough','Neighborhood']
df.columns=cols

In [40]:
df.head()

Postal Code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [41]:
df.replace("Not assigned",np.nan,inplace=True)

In [42]:
df.head()

Postal Code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [43]:
df.dropna(subset=["Borough"],axis=0,inplace=True)
df.reset_index(drop=True,inplace=True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

In [44]:
pd.set_option('display.max_rows',None)

In [45]:
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

In [46]:
df['Postal Code'].value_counts() #check to see if there are any duplicate values.

M5W    1
M4X    1
M1L    1
M4T    1
M1C    1
M1P    1
M4P    1
M9N    1
M1E    1
M1M    1
M1G    1
M3M    1
M7R    1
M9V    1
M4K    1
M8Z    1
M4G    1
M4B    1
M4N    1
M7Y    1
M9B    1
M5T    1
M5J    1
M9R    1
M6S    1
M4R    1
M6P    1
M3N    1
M5K    1
M6N    1
M1B    1
M6M    1
M3H    1
M6L    1
M5P    1
M9L    1
M4J    1
M9C    1
M3B    1
M4V    1
M5G    1
M4L    1
M2J    1
M8W    1
M1K    1
M5V    1
M8Y    1
M5R    1
M4E    1
M3A    1
M4M    1
M9A    1
M5E    1
M6A    1
M3L    1
M2M    1
M1S    1
M4C    1
M1W    1
M9P    1
M2N    1
M1N    1
M5N    1
M5C    1
M7A    1
M4H    1
M1J    1
M5L    1
M6R    1
M2R    1
M5M    1
M1T    1
M4Y    1
M6J    1
M9M    1
M2H    1
M2K    1
M1H    1
M5A    1
M5B    1
M1R    1
M5X    1
M9W    1
M5S    1
M4A    1
M1V    1
M6C    1
M2L    1
M3J    1
M6K    1
M2P    1
M6B    1
M1X    1
M6H    1
M5H    1
M3K    1
M8X    1
M4W    1
M8V    1
M6G    1
M6E    1
M3C    1
M4S    1
Name: Postal Code, dtype: int64

In [47]:
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

In [48]:
df['Neighborhood']=df['Neighborhood'].str.replace('/',',')

In [49]:
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

In [50]:
df.shape

(103, 3)

In [51]:
data='http://cocl.us/Geospatial_data'

d1=pd.read_csv(data)

In [52]:
d1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [53]:
df.head()

Postal Code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park , Harbourfront
3         M6A        North York             Lawrence Manor , Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government

In [54]:
dfll=pd.merge(df,d1,how='outer')

<strong> Data Frame with Latitude and Longitude Values</strong>

In [55]:
dfll

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

In [56]:
df_t=dfll[dfll['Borough'].str.contains("Toronto")]

df_t.reset_index(drop=True,inplace=True)

df_t

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                          Regent Park , Harbourfront  43.654260 -79.360636  
1        Queen's Park , Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                          Richmond , Adelaide , King  43.650571 -79.384568  
9                       Dufferin , Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East , Union Station , Toronto Is...  43.640816 -79.381752  
11                          Little Portugal , Trinity  43.647927 -79.419750  
12                      The Danforth West , Riverdale  43.679557 -79.352188  
13          Toronto Dominion Centre , Design Exchange  43.647177 -79.381576  
14     Brockton , Parkdale Village , Exhibition Place  43.636847 -79.428191  
15                    India Bazaar , The Beaches West  43.668999 -79.315572  
16                    Commerce Court , Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                     High Park , The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24              The Annex , North Midtown , Yorkville  43.672710 -79.405678  
25                            Parkdale , Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                    University of Toronto , Harbord  43.662696 -79.400049  
28                                Runnymede , Swansea  43.651571 -79.484450  
29                       Moore Park , Summerhill East  43.689574 -79.383160  
30        Kensington Market , Chinatown , Grange Park  43.653206 -79.400049  
31  Summerhill West , Rathnelly , South Hill , For...  43.686412 -

<strong> Explore and cluster the neighborhoods in Toronto </strong>

Laitude and longitude of torronto : 43.6532 , 79.3832

In [58]:
address = 'Toronto, TO'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
longitude = location.longitude
latitude = location.latitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [59]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_t['Latitude'], df_t['Longitude'], df_t['Borough'], df_t['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<strong> Exploring neighborhoods and segmenting them with Foursquare</strong>

In [60]:
CLIENT_ID = 'R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D' # your Foursquare ID
CLIENT_SECRET = '1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D: ' + CLIENT_ID)
print('1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5:' + CLIENT_SECRET)

Your credentails:
R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D: R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D
1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5:1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5


In [63]:
df_t.loc[0,'Neighborhood']

'Regent Park , Harbourfront'

In [64]:
df_t.head()

Postal Code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park , Harbourfront   
1         M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
2         M5B  Downtown Toronto                      Garden District, Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [65]:
neighborhood_latitude = df_t.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_t.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_t.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park , Harbourfront are 43.6542599, -79.3606359.


In [66]:
import requests

LIMIT = 100 

radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id=R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D&client_secret=1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 


'https://api.foursquare.com/v2/venues/explore?&client_id=R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D&client_secret=1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [67]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e98305529ce6a001b0e8d82'},
 'response': {'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.653446723052674,
          'lng': -79.3620167174383}],
        'distance': 143,
       

In [69]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [83]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

nearby_venues

venue.name  \
0                       Roselle Desserts   
1                          Tandem Coffee   
2                 Cooper Koo Family YMCA   
3                    Body Blitz Spa East   
4                     Morning Glory Cafe   
5                         Impact Kitchen   
6                        Corktown Common   
7                 Figs Breakfast & Lunch   
8       The Distillery Historic District   
9               Dominion Pub and Kitchen   
10              Distillery Sunday Market   
11                   SOMA chocolatemaker   
12                        Rooster Coffee   
13                              Cacao 70   
14                                  Arvo   
15                       Alumnae Theatre   
16                       Sumach Espresso   
17  Young Centre for the Performing Arts   
18                             Starbucks   
19                    The Extension Room   
20                   Brick Street Bakery   
21            Cluny Bistro & Boulangerie   
22                        Underpass Park   
23                       ODIN Cafe + Bar   
24                Parliament Square Park   
25                      Cocina Economica   
26                             El Catrin   
27           The Sweet Escape Patisserie   
28                       Berkeley Church   
29                           Caffe Furbo   
30               Dark Horse Espresso Bar   
31                       The Yoga Lounge   
32                     Mill St. Brew Pub   
33                      Greg's Ice Cream   
34                    John Fluevog Shoes   
35                    Soulpepper Theatre   
36                          Arta Gallery   
37           The Abnormal Beauty Company   
38                            The Aviary   
39                                Vistek   
40                       TD Canada Trust   
41                        The Beer Store   
42                       Savoury Grounds   
43         Residence & Conference Centre   
44                      The Healthy Road   
45                             Wine Rack   
46                                 FUEL+   
47                            GW General   

                                     venue.categories  venue.location.lat  \
0   [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...           43.653447   
1   [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.653559   
2   [{'id': '52e81612bcbc57f1066b7a37', 'name': 'D...           43.653249   
3   [{'id': '4bf58dd8d48988d1ed941735', 'name': 'S...           43.654735   
4   [{'id': '4bf58dd8d48988d143941735', 'name': 'B...           43.653947   
5   [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...           43.656369   
6   [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           43.655618   
7   [{'id': '4bf58dd8d48988d143941735', 'name': 'B...           43.655675   
8   [{'id': '4deefb944765f83613cdba6e', 'name': 'H...           43.650244   
9   [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...           43.656919   
10  [{'id': '4bf58dd8d48988d1fa941735', 'name': 'F...           43.650075   
11  [{'id': '52f2ab2ebcbc57f1066b8b31', 'name': 'C...           43.650622   
12  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.651900   
13  [{'id': '4bf58dd8d48988d1d0941735', 'name': 'D...           43.650067   
14  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.649963   
15  [{'id': '4bf58dd8d48988d137941735', 'name': 'T...           43.652756   
16  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.658135   
17  [{'id': '4bf58dd8d48988d1f2931735', 'name': 'P...           43.650825   
18  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.651613   
19  [{'id': '4bf58dd8d48988d175941735', 'name': 'G...           43.653313   
20  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...           43.650574   
21  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...           43.650565   
22  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           43.655764   
23  [{'id': '4bf58dd8d489

In [84]:
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149

In [85]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


## <strong> Exploring neighborhoods in Toronto</strong>

In [87]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id=R4ZBTQK1BP10AQGANFHH5TQKMPMBN5J3R040XBHJSWQZL35D&client_secret=1NYHEFZORGORZFCJOLLRUJBZV51XIW3OJTGM4HJRROFQMAQ5&v=20180605&ll=43.6534817,-79.3839347&radius=500&limit=100'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
        
      # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [88]:
Toronto_venues = getNearbyVenues(names=df_t['Neighborhood'],
                                   latitudes=df_t['Latitude'],
                                   longitudes=df_t['Longitude']
                                  )

Regent Park , Harbourfront
Queen's Park , Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond , Adelaide , King
Dufferin , Dovercourt Village
Harbourfront East , Union Station , Toronto Islands
Little Portugal , Trinity
The Danforth West , Riverdale
Toronto Dominion Centre , Design Exchange
Brockton , Parkdale Village , Exhibition Place
India Bazaar , The Beaches West
Commerce Court , Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park , The Junction South
North Toronto West
The Annex , North Midtown , Yorkville
Parkdale , Roncesvalles
Davisville
University of Toronto , Harbord
Runnymede , Swansea
Moore Park , Summerhill East
Kensington Market , Chinatown , Grange Park
Summerhill West , Rathnelly , South Hill , Forest Hill SE , Deer Park
CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst  Quay , South Niagara , Island airport
Rosed

In [89]:
print(Toronto_venues.shape)
Toronto_venues.head()

(3042, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park , Harbourfront               43.65426              -79.360636   
1  Regent Park , Harbourfront               43.65426              -79.360636   
2  Regent Park , Harbourfront               43.65426              -79.360636   
3  Regent Park , Harbourfront               43.65426              -79.360636   
4  Regent Park , Harbourfront               43.65426              -79.360636   

                               Venue  Venue Latitude  Venue Longitude  \
0                   Downtown Toronto       43.653232       -79.385296   
1             Nathan Phillips Square       43.652270       -79.383516   
2  Eggspectation Bell Trinity Square       43.653144       -79.381980   
3                            Japango       43.655268       -79.385165   
4                             Indigo       43.653515       -79.380696   

     Venue Category  
0      Neighborhood  
1             Plaza  
2    Breakfast Spot  
3  Sushi Restaurant  
4         Bookstore

In [90]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            78   
Brockton , Parkdale Village , Exhibition Place                         78   
Business reply mail Processing CentrE                                  78   
CN Tower , King and Spadina , Railway Lands , H...                     78   
Central Bay Street                                                     78   
Christie                                                               78   
Church and Wellesley                                                   78   
Commerce Court , Victoria Hotel                                        78   
Davisville                                                             78   
Davisville North                                                       78   
Dufferin , Dovercourt Village                                          78   
First Canadian Place , Underground city                                78   
Forest Hill North & West                                               78   
Garden District, Ryerson                                               78   
Harbourfront East , Union Station , Toronto Isl...                     78   
High Park , The Junction South                                         78   
India Bazaar , The Beaches West                                        78   
Kensington Market , Chinatown , Grange Park                            78   
Lawrence Park                                                          78   
Little Portugal , Trinity                                              78   
Moore Park , Summerhill East                                           78   
North Toronto West                                                     78   
Parkdale , Roncesvalles                                                78   
Queen's Park , Ontario Provincial Government                           78   
Regent Park , Harbourfront                                             78   
Richmond , Adelaide , King                                             78   
Rosedale                                                               78   
Roselawn                                                               78   
Runnymede , Swansea                                                    78   
St. James Town                                                         78   
St. James Town , Cabbagetown                                           78   
Stn A PO Boxes                                                         78   
Studio District                                                        78   
Summerhill West , Rathnelly , South Hill , Fore...                     78   
The Annex , North Midtown , Yorkville                                  78   
The Beaches                                                            78   
The Danforth West , Riverdale                                          78   
Toronto Dominion Centre , Design Exchange                              78   
University of Toronto , Harbord                                        78   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             78   
Brockton , Parkdale Village , Exhibition Place                          78   
Business reply mail Processing CentrE                                   78   
CN Tower , King and Spadina , Railway Lands , H...                      78   
Central Bay Street                                                      78   
Christie                                                                78   
Church and Wellesley                                                    78   
Commerce Court , Victoria Hotel                                         78   
Davisville                                                              78   
Davisville North                    

In [91]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 59 uniques categories.


In [92]:
# one hot encoding
toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Women's Store  American Restaurant  Asian Restaurant  Bakery  Bank  Bar  \
0              0                    0                 0       0     0    0   
1              0                    0                 0       0     0    0   
2              0                    0                 0       0     0    0   
3              0                    0                 0       0     0    0   
4              0                    0                 0       0     0    0   

   Beer Bar  Bookstore  Breakfast Spot  Bubble Tea Shop  Café  \
0         0          0               0                0     0   
1         0          0               0                0     0   
2         0          0               1                0     0   
3         0          0               0                0     0   
4         0          1               0                0     0   

   Chinese Restaurant  Clothing Store  Cocktail Bar  Coffee Shop  \
0                   0               0             0            0   
1                   0               0             0            0   
2                   0               0             0            0   
3                   0               0             0            0   
4                   0               0             0            0   

   Colombian Restaurant  Comic Shop  Concert Hall  Cosmetics Shop  \
0                     0           0             0               0   
1                     0           0             0               0   
2                     0           0             0               0   
3                     0           0             0               0   
4                     0           0             0               0   

   Department Store  Dessert Shop  Diner  Electronics Store  \
0                 0             0      0                  0   
1                 0             0      0                  0   
2                 0             0      0                  0   
3                 0             0      0                  0   
4                 0             0      0                  0   

   Fast Food Restaurant  Food & Drink Shop  Food Court  \
0                     0                  0           0   
1                     0                  0           0   
2                     0                  0           0   
3                     0                  0           0   
4                     0                  0           0   

   Furniture / Home Store  Gastropub  General Travel  Gym  \
0                       0          0               0    0   
1                       0          0               0    0   
2                       0          0               0    0   
3                       0          0               0    0   
4                       0          0               0    0   

   Gym / Fitness Center  Hotel  Ice Cream Shop  Japanese Restaurant  \
0                     0      0               0                    0   
1                     0      0               0                    0   
2                     0      0               0                    0   
3                     0      0               0                    0   
4                     0      0               0                    0   

   Juice Bar  Latin American Restaurant  Mexican Restaurant  \
0          0                          0                   0   
1          0                          0                   0   
2          0                          0                   0   
3          0                          0                   0   
4          0                          0                   0   

   Middle Eastern Restaurant  Miscellaneous Shop  Modern European Restaurant  \
0                          0                   0                           0   
1                          0                   0                           0   
2                          0                   0                           0   
3                          0                   0                           0   
4                 

In [93]:
toronto_onehot.shape

(3042, 59)

In [94]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Women's Store  \
0                                         Berczy Park       0.012821   
1      Brockton , Parkdale Village , Exhibition Place       0.012821   
2               Business reply mail Processing CentrE       0.012821   
3   CN Tower , King and Spadina , Railway Lands , ...       0.012821   
4                                  Central Bay Street       0.012821   
5                                            Christie       0.012821   
6                                Church and Wellesley       0.012821   
7                     Commerce Court , Victoria Hotel       0.012821   
8                                          Davisville       0.012821   
9                                    Davisville North       0.012821   
10                      Dufferin , Dovercourt Village       0.012821   
11            First Canadian Place , Underground city       0.012821   
12                           Forest Hill North & West       0.012821   
13                           Garden District, Ryerson       0.012821   
14  Harbourfront East , Union Station , Toronto Is...       0.012821   
15                     High Park , The Junction South       0.012821   
16                    India Bazaar , The Beaches West       0.012821   
17        Kensington Market , Chinatown , Grange Park       0.012821   
18                                      Lawrence Park       0.012821   
19                          Little Portugal , Trinity       0.012821   
20                       Moore Park , Summerhill East       0.012821   
21                                 North Toronto West       0.012821   
22                            Parkdale , Roncesvalles       0.012821   
23       Queen's Park , Ontario Provincial Government       0.012821   
24                         Regent Park , Harbourfront       0.012821   
25                         Richmond , Adelaide , King       0.012821   
26                                           Rosedale       0.012821   
27                                           Roselawn       0.012821   
28                                Runnymede , Swansea       0.012821   
29                                     St. James Town       0.012821   
30                       St. James Town , Cabbagetown       0.012821   
31                                     Stn A PO Boxes       0.012821   
32                                    Studio District       0.012821   
33  Summerhill West , Rathnelly , South Hill , For...       0.012821   
34              The Annex , North Midtown , Yorkville       0.012821   
35                                        The Beaches       0.012821   
36                      The Danforth West , Riverdale       0.012821   
37          Toronto Dominion Centre , Design Exchange       0.012821   
38                    University of Toronto , Harbord       0.012821   

    American Restaurant  Asian Restaurant    Bakery      Bank       Bar  \
0              0.025641          0.012821  0.012821  0.012821  0.012821   
1              0.025641          0.012821  0.012821  0.012821  0.012821   
2              0.025641          0.012821  0.012821  0.012821  0.012821   
3              0.025641          0.012821  0.012821  0.012821  0.012821   
4              0.025641          0.012821  0.012821  0.012821  0.012821   
5              0.025641          0.012821  0.012821  0.012821  0.012821   
6              0.025641          0.012821  0.012821  0.012821  0.012821   
7              0.025641          0.012821  0.012821  0.012821  0.012821   
8              0.025641          0.012821  0.012821  0.012821  0.012821   
9              0.025641          0.012821  0.012821  0.012821  0.012821   
10             0.025641          0.012821  0.012821  0.012821  0.012821   
11             0.025641          0.012821  0.012821  0.012821  0.012821   
12             0.025641          0.012821  0.012821  0.012821  0.012821   
13             0.025641          0.012821  0.012821  0.012821  0.012821   
14             0.025641          0.

In [95]:
toronto_grouped.shape

(39, 59)

In [96]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0  Clothing Store  0.06
1     Coffee Shop  0.05
2      Restaurant  0.04
3           Plaza  0.04
4  Breakfast Spot  0.03


----Brockton , Parkdale Village , Exhibition Place----
            venue  freq
0  Clothing Store  0.06
1     Coffee Shop  0.05
2      Restaurant  0.04
3           Plaza  0.04
4  Breakfast Spot  0.03


----Business reply mail Processing CentrE----
            venue  freq
0  Clothing Store  0.06
1     Coffee Shop  0.05
2      Restaurant  0.04
3           Plaza  0.04
4  Breakfast Spot  0.03


----CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst  Quay , South Niagara , Island airport----
            venue  freq
0  Clothing Store  0.06
1     Coffee Shop  0.05
2      Restaurant  0.04
3           Plaza  0.04
4  Breakfast Spot  0.03


----Central Bay Street----
            venue  freq
0  Clothing Store  0.06
1     Coffee Shop  0.05
2      Restaurant  0.04
3           Plaza  0.04
4  Breakfast Spot  0.03




In [98]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [99]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park        Clothing Store   
1     Brockton , Parkdale Village , Exhibition Place        Clothing Store   
2              Business reply mail Processing CentrE        Clothing Store   
3  CN Tower , King and Spadina , Railway Lands , ...        Clothing Store   
4                                 Central Bay Street        Clothing Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop            Restaurant                 Plaza   
1           Coffee Shop            Restaurant                 Plaza   
2           Coffee Shop            Restaurant                 Plaza   
3           Coffee Shop            Restaurant                 Plaza   
4           Coffee Shop            Restaurant                 Plaza   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store                 Diner                 Hotel   
1     Electronics Store                 Diner                 Hotel   
2     Electronics Store                 Diner                 Hotel   
3     Electronics Store                 Diner                 Hotel   
4     Electronics Store                 Diner                 Hotel   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Steakhouse               Theater         Breakfast Spot  
1            Steakhouse               Theater         Breakfast Spot  
2            Steakhouse               Theater         Breakfast Spot  
3            Steakhouse               Theater         Breakfast Spot  
4            Steakhouse               Theater         Breakfast Spot

## <strong> Clustering using K means </strong>

In [102]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/cluster/k_means_.py:971: ConvergenceWarning: Number of distinct clusters (1) found smaller than n_clusters (5). Possibly due to duplicate points in X.
  return_n_iter=True)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [103]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged =df_t

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park , Harbourfront   
1         M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
2         M5B  Downtown Toronto                      Garden District, Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0        Clothing Store   
1  43.662301 -79.389494               0        Clothing Store   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0        Clothing Store   
4  43.676357 -79.293031               0        Clothing Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop            Restaurant                 Plaza   
1           Coffee Shop            Restaurant                 Plaza   
2           Coffee Shop            Restaurant                 Plaza   
3           Coffee Shop            Restaurant                 Plaza   
4           Coffee Shop            Restaurant                 Plaza   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store                 Diner                 Hotel   
1     Electronics Store                 Diner                 Hotel   
2     Electronics Store                 Diner                 Hotel   
3     Electronics Store                 Diner                 Hotel   
4     Electronics Store                 Diner                 Hotel   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Steakhouse               Theater         Breakfast Spot  
1            Steakhouse               Theater         Breakfast Spot  
2            Steakhouse               Theater         Breakfast Spot  
3            Steakhouse               Theater         Breakfast Spot  
4            Steakhouse               Theater         Breakfast Spot

In [104]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<strong> Examining clusters </strong>

In [107]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0        Clothing Store   
1   Downtown Toronto               0        Clothing Store   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0        Clothing Store   
4       East Toronto               0        Clothing Store   
5   Downtown Toronto               0        Clothing Store   
6   Downtown Toronto               0        Clothing Store   
7   Downtown Toronto               0        Clothing Store   
8   Downtown Toronto               0        Clothing Store   
9       West Toronto               0        Clothing Store   
10  Downtown Toronto               0        Clothing Store   
11      West Toronto               0        Clothing Store   
12      East Toronto               0        Clothing Store   
13  Downtown Toronto               0        Clothing Store   
14      West Toronto               0        Clothing Store   
15      East Toronto               0        Clothing Store   
16  Downtown Toronto               0        Clothing Store   
17      East Toronto               0        Clothing Store   
18   Central Toronto               0        Clothing Store   
19   Central Toronto               0        Clothing Store   
20   Central Toronto               0        Clothing Store   
21   Central Toronto               0        Clothing Store   
22      West Toronto               0        Clothing Store   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0        Clothing Store   
25      West Toronto               0        Clothing Store   
26   Central Toronto               0        Clothing Store   
27  Downtown Toronto               0        Clothing Store   
28      West Toronto               0        Clothing Store   
29   Central Toronto               0        Clothing Store   
30  Downtown Toronto               0        Clothing Store   
31   Central Toronto               0        Clothing Store   
32  Downtown Toronto               0        Clothing Store   
33  Downtown Toronto               0        Clothing Store   
34  Downtown Toronto               0        Clothing Store   
35  Downtown Toronto               0        Clothing Store   
36  Downtown Toronto               0        Clothing Store   
37  Downtown Toronto               0        Clothing Store   
38      East Toronto               0        Clothing Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Coffee Shop            Restaurant                 Plaza   
1            Coffee Shop            Restaurant                 Plaza   
2            Coffee Shop            Restaurant                 Plaza   
3            Coffee Shop            Restaurant                 Plaza   
4            Coffee Shop            Restaurant                 Plaza   
5            Coffee Shop            Restaurant                 Plaza   
6            Coffee Shop            Restaurant                 Plaza   
7            Coffee Shop            Restaurant                 Plaza   
8            Coffee Shop            Restaurant                 Plaza   
9            Coffee Shop            Restaurant                 Plaza   
10           Coffee Shop            Restaurant                 Plaza   
11           Coffee Shop            Restaurant                 Plaza   
12           Coffee Shop            Restaurant                 Plaza   
13           Coffee Shop            Restaurant                 Plaza   
14           Coffee Shop            Restaurant                 Plaza   
15           Coffee Shop            Restaurant                 Plaza   
16           Coffee Shop            Restaurant                 Plaza   
17           Coffee Shop            Restaurant                 Plaza   
18           Coffee Shop            Restaurant                 Plaza   
19           Coffee Shop            Restaurant                 Plaza   
20           Coffee 